drivers: wifi: Add DHCP/Static IP Support for ESP wifi driver

ExtremeGTX opened the pull request:
Add ESP DHCP/StaticIP Support including kconfig entries.

#### Other minor fixes included:
- ~~Print a warning message on shell when user executes `net ping xxx.xxx.xxx.xxx` in case of offloaded driver is being used.~~
- ~~Rename kconfig prompt `add support for Wi-Fi Drivers` to `Wi-Fi Drivers`~~

jukkar commented on drivers/wifi/esp/esp.c:
```diff
@@ -720,6 +725,16 @@ static void esp_init_work(struct k_work *work)
+		/* enable Static IP Config */
+		SETUP_CMD_NOHANDLE(ESP_CMD_DHCP_ENABLE(1, 0)),
+		SETUP_CMD_NOHANDLE(ESP_CMD_SET_IP(CONFIG_WIFI_ESP_IP_ADDRESS,
+										  CONFIG_WIFI_ESP_IP_GATEWAY,
+										  CONFIG_WIFI_ESP_IP_MASK)),
```
Indentation is off here. Please make sure that your tab length is 8

jukkar reviewed (commented):
Looks ok, minor nit about indentation.

WilliamGFish commented:
I'll give this test later this week for you, on the modified Particle Argon board.

mniestroj commented on drivers/wifi/esp/esp.c:
```diff
@@ -257,7 +257,11 @@ static void esp_ip_addr_work(struct k_work *work)
 	/* update interface addresses */
 	net_if_ipv4_set_gw(dev->net_iface, &dev->gw);
 	net_if_ipv4_set_netmask(dev->net_iface, &dev->nm);
+#if WIFI_ESP_IP_STATIC
```
I believe it should be `CONFIG_WIFI_ESP_IP_STATIC` instead.

mniestroj commented on drivers/wifi/esp/esp.c:
```diff
@@ -710,6 +714,7 @@ static void esp_init_work(struct k_work *work)
 	struct esp_data *dev;
 	int ret;
 	static struct setup_cmd setup_cmds[] = {
+
```
No need to add that NL here.

mniestroj commented on drivers/wifi/esp/esp.c:
```diff
@@ -720,6 +725,16 @@ static void esp_init_work(struct k_work *work)
+						  CONFIG_WIFI_ESP_IP_MASK)),
+#else
+		/* enable DHCP */
+		SETUP_CMD_NOHANDLE(ESP_CMD_DHCP_ENABLE(1, 1)),
```
Should we add `ESP_MODE_SOFTAP` and `ESP_MODE_STATION` macros, so above line will get more readable:
```
SETUP_CMD_NOHANDLE(ESP_CMD_DHCP_ENABLE(ESP_MODE_STATION, 1)),
```
?

mniestroj reviewed (commented):
I guess `drivers: wifi: rename WiFi kconfig prompt` and `net: ping: add warning if ping is not supported` could go with separate PRs, so they get into master regardless of ESP WiFi changes.

ExtremeGTX commented:
@mniestroj thanks for the review.
I have done all the requested changes, also a separate PR is created for each of the commits.

mniestroj commented on drivers/wifi/esp/esp.h:
```diff
@@ -95,6 +95,25 @@ extern "C" {
+#define ESP_MODE_SOFTAP		0
+#else
+#define ESP_MODE_SOFTAP		2
+#endif
```
I've looked into docs if those esp mode values are shared across different commands. Unfortunately that is not true for `CONFIG_WIFI_ESP_AT_VERSION_1_7` case, as `AT+CWMODE_CUR` assumes SoftAP mode to be represented by value `2`, but for `AT+CWMODE_CUR` command by value `0` (as you've implemented above).

So either we need those `ESP_MODE_*` macros to be renamed to `ESP_CWDHCP_MODE_*` or somehow map `ESP_MODE_*` to resolve to different values depending on command in which they are used. The former approach seems easier, but more verbose. For now I don't know how to implement the latter...

ExtremeGTX replied:
> Some few more things in comments. Additionally I don't think `- Add ESP DHCP Support` is the correct sentence in git commit message, as well as `esp DHCP/Static IP Support` in commit title. We already relied on DHCP to be working. The only thing that changes now is that you force DHCP by calling `AT+CWDHCP` command, instead of relying it to be the default in ESP firmware.

On firmware v1.7.4 this never worked for me on esp8266-01, I tried also to do `AT+RESTORE` then connect, but either ESP fails to get an IP Address and disconnect or it got some weird IP starts with `252.xxx.xxx.xxx`

Also, I think it is better to get things under control from Zephyr side, so we make sure that our AT commands sequence are working and we don't rely on ESP init sequence which might change.

ExtremeGTX replied:
Good catch ! 👍
Will think about it

ExtremeGTX replied:
Solved as agreed on Slack
Thanks for your support.

mniestroj reviewed (commented):
Some few more things in comments. Additionally I don't think `- Add ESP DHCP Support` is the correct sentence in git commit message, as well as `esp DHCP/Static IP Support` in commit title. We already relied on DHCP to be working. The only thing that changes now is that you force DHCP by calling `AT+CWDHCP` command, instead of relying it to be the default in ESP firmware.

ExtremeGTX commented:
> Some few more things in comments. Additionally I don't think `- Add ESP DHCP Support` is the correct sentence in git commit message, as well as `esp DHCP/Static IP Support` in commit title. We already relied on DHCP to be working. The only thing that changes now is that you force DHCP by calling `AT+CWDHCP` command, instead of relying it to be the default in ESP firmware.

On firmware v1.7.4 this never worked for me on esp8266-01, I tried also to do AT+RESTORE then connect, but either ESP fails to get an IP Address and disconnect or it got some weird IP starts with 252.xxx.xxx.xxx

Also, I think it is better to get things under control from Zephyr side, so we make sure that our AT commands sequence are working and we don't rely on ESP init sequence which might change.

mniestroj commented on drivers/wifi/esp/Kconfig.esp:
```diff
@@ -80,6 +80,37 @@ config WIFI_ESP_RX_NET_PKT_ALLOC_TIMEOUT
 	default 5000
 	help
 	  Network interface RX net_pkt allocation timeout in milliseconds.
+choice
```
Please add one NL before `choice`.

mniestroj reviewed (commented):
Just one minor nit, otherwise looks good.

ExtremeGTX committed:
drivers: wifi: esp DHCP/Static IP Support

- Add ESP DHCP Support
- Add ESP Static IP Support including
 KConfig entries for configuring IP,Gateway and netMask

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

WilliamGFish commented:
> @WilliamGFish did you get the ESP32 on the Argon to work?

Just a thought, to get the Argon working correctly will need to change the firmware to a standard version as Particle have added their own which restricts AT command set.

Try: https://learn.adafruit.com/adding-a-wifi-co-processor-to-circuitpython-esp8266-esp32/program-particle-argon 

Billy..
